### PR TITLE
Fix / Terms links with bundle quotes

### PR DESCRIPTION
**Pull request by styrelius**
## What?

The list of Terms & Conditions links in checkout is fixed so that it actually contains all links from all quotes:
- All logic of getting the `insuranceTerms` from the quotes into a desirable shape is extracted to its own function `getInsuranceTerms`, in `InsuranceSummaryTermsLinks.tsx`.
- A bunch of tests describing the `getInsuranceTerms` function are added to new test file `InsuranceSummaryTermsLinks.test.ts`
- The list of terms is now also sorted in the following order: 
  1. `TERMS_AND_CONDITIONS` type term from main quote
  2. other `TERMS_AND_CONDITION` type terms (if it's a bundle)
  3. all other terms (still excluding `PRIVACY_POLICY` term types though).

## Why?

The current implementation which luckily enough is not yet in production yet since I realised that it wasn't working properly will only display terms links from the main quote (i.e. Home Content quote in Norway, in the future also Home Content quote in Denmark).

_Referenced ticket(s): [GRW-265]_

### 🇸🇪 Sweden
2021-04-20-terms-links-se

### 🇳🇴 Norway
**Kombo:**
2021-04-20-terms-links-no-combo

**Innbo:**
2021-04-20-terms-links-no-innbo

**Reise:**
2021-04-20-terms-links-no-travel

**Bundle**
_Unfortunately the Travel insurance terms and Home Content insurance terms have the same `displayName` in English (which we get from back-end), so it looks like a mistake here but it's not. Will ask for it to be changed on the back-end._
2021-04-20-terms-links-no-en

### 🇩🇰 Denmark
_still placeholder data obvi_
2021-04-20-terms-links-dk

[GRW-265]: https://hedvig.atlassian.net/browse/GRW-265

### CODE REVIEW

**Comment by gustaveen** [comment]
Well fought battle against T&Cs! @styrelius 🙌  Looks really nice! Still waiting for staging to be back up to test it, but some initial initial thoughts and questions. 

Isn't pre-sale information still unique for each contract (just looking at the terms page https://www.hedvig.com/no/terms). I know we just link to the terms page temporarily but we should still list that for every quote right? And that might mean an update to the title like `Pre-sale information travel` for example?

I know our goal is to have this neatly structured for the member. To me it's quite hard to follow the logic on what to show or not though 😅  I'd like us to have a simpler solution where we just render what we get from BE. For example I don't think Privacy policy should be a part of the `termsAndConditions` object but rather it's own field on the quote. That would simplify it a bit at least.  So that would just leave us with filtering out duplicates of general terms. But it might be hard to change now ofc.

Another idea is to just output the terms for one quote at the time. You could probably do it a bit neater than my example though 😄  
Screenshot 2021-04-21 at 15 33 59

**Comment by styrelius** [comment]
@gustaveen The pre-sale information is not unique for each quote **currently** since we (temporarily) get the same url from the back-end for both quotes in the bundle. The filtering of duplicates is made on the basis of having _both_ the same `displayName` _and_ `url`. But yes, once we'll get different url's from BE for pre-sale information that `displayName` will be rendered twice if we don't ask the back-end to kindly change that `displayName` too 🤔 

Since they probably don't want to change the structure of the quote on the back-end I think we'll have to do the removing of the privacy policy ourselves.

I did think about your solution as well but then we'll have the `General Terms` duplicated which is annoying. I'd really like us to have as few links as possible to not... overwhelm the user you know. 

I understand the logic can be hard to follow, that's partly why I wrote the tests.

File: src/client/pages/OfferNew/Checkout/InsuranceSummaryTermsLinks.tsx [hunk 1]
```diff
@@ -38,27 +39,96 @@ const getUrl = ({ currentLocale, termType, urlFromBackend }: GetUrlParams) => {
...
 
-export const InsuranceSummaryTermsLinks: React.FC<Props> = ({ mainQuote }) => {
-  const currentLocale = useCurrentLocale()
+export type Terms = Term[]
 
-  const { insuranceTerms } = mainQuote
+type GetSortedParams = {
+  termsArray: Terms
+  offerData: OfferData
+}
```
[the remark this answers is not in the document]
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/InsuranceSummaryTermsLinks.tsx:53]: The `Terms` type is used in the test file but maybe you still think we should use `Term[]` in both files?

File: src/client/pages/OfferNew/Checkout/InsuranceSummaryTermsLinks.tsx [hunk 2]
```diff
@@ -38,27 +39,96 @@ const getUrl = ({ currentLocale, termType, urlFromBackend }: GetUrlParams) => {
...
+    ({ termType }) => termType !== InsuranceTermType.TermsAndConditions,
+  )
+  return [...termsAndConditionsSorted, ...restOfTerms]
+}
+
+type GetInsuranceTermsParams = {
+  offerData: OfferData
+}
+
+export const getInsuranceTerms = ({ offerData }: GetInsuranceTermsParams) => {
```
[the remark this answers is not in the document]
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/InsuranceSummaryTermsLinks.tsx:77]: Haha yes! 👌  I guess I had more parameters initially and added the type alias for that reason 🙈

[the remark this answers is not in the document]
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/InsuranceSummaryTermsLinks.tsx:77]: No OK, I think actually the deal was that it's a bit clearer with a named parameter when this is used in the test file, and this got kind of ugly: 
image

File: src/client/pages/OfferNew/Checkout/InsuranceSummaryTermsLinks.tsx [hunk 3]
```diff
@@ -38,27 +39,96 @@ const getUrl = ({ currentLocale, termType, urlFromBackend }: GetUrlParams) => {
...
+  const allQuotesInsuranceTerms = offerData.quotes
+    .map((quote) => {
+      const quoteId = quote.id
+      const { insuranceTerms } = quote
+      return [...insuranceTerms].map(([termType, data]) => {
+        return { quoteId, termType, data }
+      })
+    })
+    .reduce((acc, cur) => {
+      return [...acc, ...cur]
```
[the remark this answers is not in the document]
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/InsuranceSummaryTermsLinks.tsx:87]: Actually this `cur` is an array of terms objects (lol this is so messy) but YES, I'll name it better!

File: src/client/pages/OfferNew/Checkout/InsuranceSummaryTermsLinks.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (gustaveen) [reply src/client/pages/OfferNew/Checkout/InsuranceSummaryTermsLinks.tsx]: I see your point 🙌   But couldn't that be said for all functions where we pass in offerData? :) I think it's nicer to just hover the the function and you will see what the params are. To me this feels a bit bloated 🙃
Screenshot 2021-04-23 at 09 08 36

[the remark this answers is not in the document]
Comment (styrelius) [reply src/client/pages/OfferNew/Checkout/InsuranceSummaryTermsLinks.tsx]: In other places where we have `offerData` as a parameter we pass in the/a variable called `offerData`. In this test file though we have two different variables with "offer data", thus it's a bit less clear by just looking at the function calls what e.g. `noCombo` is supposed to be. But if you think it's important I can change it.

**Review by gustaveen — APPROVED** [review]
Woho, let's go! 
giphy (2)
